File: Residential_Controller.py
# defining my kind of blueprint
class Column():
    def __init__(self, floors, elevators):
        self.floors = floors
        self.elevators = elevators
        self.arrayButtons = []
        self.arrayelevators = []
        for i in range(elevators):
            self.arrayelevators.append(Elevator(0, floors))
        for i in range(floors):
            if i == 0:
                self.arrayButtons.append(ExtButton(i, "up"))
            else:
                self.arrayButtons.append(ExtButton(i, "up"))
                self.arrayButtons.append(ExtButton(i, "down"))

    def findElevator(self, requestedFloor, direction):

        bestDiff = self.floors
        chosenElevator = None

        for elevator in self.arrayelevators:
            if elevator.direction == "up" and direction == "up" and requestedFloor < elevator.currentFloor:
                chosenElevator = elevator
            elif elevator.direction == "down" and direction == "down" and requestedFloor > elevator.currentFloor:
                chosenElevator = elevator
            elif elevator.status == "idle":
                chosenElevator = elevator
            else:
                for elevator in self.arrayelevators:
                    diff = abs(elevator.currentFloor - requestedFloor)
                    if diff < bestDiff:
                        chosenElevator = elevator
                        bestDiff = diff
        print("Best elevator on floor " + str(chosenElevator.currentFloor))
        return chosenElevator

    def requestElevator(self, requestedFloor, direction):
        print("Called an elevator to the floor " + str(requestedFloor))

        elevator = self.findElevator(requestedFloor, direction)

        elevator.addToQueue(requestedFloor)
        elevator.move()
        return elevator

    def requestFloor(self, elevator, requestedFloor):
        print("Moving elevator on floor " + str(elevator.currentFloor) + " to the floor " + str(requestedFloor))
        elevator.addToQueue(requestedFloor)
        elevator.closeDoors()
        elevator.move()

class Elevator():
    def __init__(self, currentFloor, floors):
        self.direction =  None
        self.floors = floors
        self.currentFloor = currentFloor
        self.status = "idle"
        self.queue = []
        self.IntButtonsList = []
        self.door = "closed"

        for i in range(self.floors):
            self.IntButtonsList.append(IntButton(i))

    def addToQueue(self, requestedFloor):
        self.queue.append(requestedFloor)

        if self.direction == "up":
            self.queue.sort(reverse=True)
        if self.direction == "down":
            self.queue.sort(reverse=True)

        print("Added floor " + str(requestedFloor) + " elevators driving passengers on " + ', '.join(str(x) for x in self.queue))

    def move(self):
        print ("Moving elevator")
        #while len function to returns the number 
        while len(self.queue) > 0:

            operate = self.queue[0]

            if self.door == "open":
                print("within 5 seconds if passage is not obstructed")
                self.closeDoors()
            if operate == self.currentFloor:
                del self.queue[0]
                self.openDoors()
            elif operate > self.currentFloor:
                self.status = "moving"
                self.direction = "up"
                self.moveUp()
            elif operate < self.currentFloor:
                self.status = "moving"
                self.direction = "down"
                self.moveDown()
        print("within 5 seconds if passage is not obstructed")
        self.closeDoors()
        print("Elevator is now idle")
        self.status = "idle"

    def moveUp(self):
        self.currentFloor += 1
        print("moving up Elevator on floor " + str(self.currentFloor))

    def moveDown(self):
        self.currentFloor -= 1
        print("moving down Elevator on floor " + str(self.currentFloor))

    def openDoors(self):
        self.door = "open"
        print("open door on floor ")

    def closeDoors(self):
        self.door="closed"
        print("close door on floor ")

class ExtButton():
    def __init__(self, requestFloor, direction):
        self.requestFloor = requestFloor
        self.direction = direction

class IntButton():
    def __init__(self, floor):
        self.floor = floor


# testing #


def test_requestElevator():
                column1 = Column(10, 2)
                column1.arrayelevators[0].currentFloor = 6
                column1.arrayelevators[0].direction  =  "up"
                column1.arrayelevators[0].status =  "idle"
                column1.arrayelevators[0].queue = [1, 5, 8]
                column1.arrayelevators[1].currentFloor = 3
                column1.arrayelevators[1].direction  =  "up"
                column1.arrayelevators[1].status =  "moving"
                column1.arrayelevators[1].queue = [2, 1]
                
                column1.requestElevator(4, "down")
                column1.requestElevator(9, "down")

test_requestElevator()

File: Residential_Controller.js
//The  Column function contructor has three methods from my column are :- findElevator, requestElevator and requestFloor
class Column {
	constructor(floors, elevators) {
		this.floors = floors;
		this.elevators = elevators;
		this.arrayExtButton = [];
		this.arrayElevators = [];
		for (let i = 0; i < elevators; i++) {
			this.arrayElevators.push(new Elevator(0, floors));
		}
		for (let i = 0; i < floors; i++) {
			if (i == 0) {
				this.arrayExtButton.push(new ExtButton(i, "up", false));
			} else {
				this.arrayExtButton.push(new ExtButton(i, "up", false));
				this.arrayExtButton.push(new ExtButton(i, "down", false));
			}
		}
	}
    
	findElevator(requestedFloor, direction) {
		let selectedElevator = null;
		let bestDiff = this.floors;

		for (let i = 0; i < this.arrayElevators.length; i++) {
			if (this.arrayElevators[i].direction === "up" && direction === "up" && requestedFloor > this.arrayElevators[i].currentFloor) {
				selectedElevator = this.arrayElevators[i];
			} else if (this.arrayElevators[i].direction === "down" && direction === "down" && requestedFloor < this.arrayElevators[i].currentFloor) {
				selectedElevator = this.arrayElevators[i];
			} else if (this.arrayElevators[i].status === "idle") {
				selectedElevator = this.arrayElevators[i];
			} else {
				for (let i = 0; i < this.arrayElevators.length; i++) {
					let diff = Math.abs(this.arrayElevators[i].currentFloor - requestedFloor);
					if (diff < bestDiff) {
						selectedElevator = this.arrayElevators[i];
						bestDiff = diff;
					}
				}
			}
		}
		console.log("System has find the best elevator on floor " + selectedElevator.currentFloor);
		return selectedElevator;
	}

	requestElevator(requestedFloor, direction) {
		if (requestedFloor > this.floors) return console.log("Floor " + requestedFloor + " doesn't exist!");
		console.log("Call for an elevator to the floor " + requestedFloor);

		let elevator = this.findElevator(requestedFloor, direction);

		elevator.addToQueue(requestedFloor);
		elevator.move();
		//console.log(elevator);
		return elevator;
	}
	
	requestFloor(elevator, requestedFloor) {
		console.log("Moving elevator on floor " + elevator.currentFloor + " to the floor " + requestedFloor);
		elevator.addToQueue(requestedFloor);
		elevator.closeDoors();
		elevator.move();
	}
	
}

class Elevator {
	constructor(currentFloor, floors) {

		this.direction = null;
		this.floors = floors;
		this.currentFloor = currentFloor;
		this.status = "idle";
		this.queue = [];
		this.InButtonsList = [];
		this.door = "closed";

		for (let i = 0; i < this.floors; i++) {
			this.InButtonsList.push(new InButton(i));
		}
	}
	
	addToQueue(requestedFloor) {
		this.queue.push(requestedFloor)

		if (this.direction == "up") {
			this.queue.sort((a, b) => a - b)
		}
		if (this.direction == "down") {
			this.queue.sort((a, b) => b - a)
		}
		//console.log(this.queue.join(", "));
		console.log("Added floor " + requestedFloor + " to the elevator's queue which is now:  " + this.queue.join(", "));
	}
	
	move() {
		//console.log("Move elevator");
		while (this.queue.length > 0) {

			let operate = this.queue[0];

			if (this.door === "open") {
				console.log("Within 5 seconds the door closes if passage is not obstructed");
				this.closeDoors();
			}
			if (operate === this.currentFloor) {
				this.queue.shift();
				this.openDoors();
			}
			if (operate > this.currentFloor) {
				this.status = "moving";
				this.direction = "up";
				this.moveUp();
			}
			if (operate < this.currentFloor) {
				this.status = "moving";
				this.direction = "down";
				this.moveDown()
			}
		}
		console.log("Within 5 seconds door the closes if passage is not obstructed");
		this.closeDoors();
		console.log("Elevator is now idle on floor " + this.currentFloor);
		this.status = "is idle";
	}
	
	moveUp() {
		this.currentFloor++;
		console.log("Moving up Elevator on floor " + this.currentFloor);
	}

	moveDown() {
		this.currentFloor--;
		console.log("Moving down Elevator on floor " + this.currentFloor);
	}

	openDoors() {
		this.door = "open"
		console.log("Open door on floor " + this.currentFloor);
	}

	closeDoors() {
		this.door = "closed"
		console.log("Close door on floor " + this.currentFloor);
	}

}

class ExtButton {
	constructor(requestFloor, direction) {
		this.requestFloor = requestFloor;
		this.direction = direction;
	}
}

class InButton {
	constructor(floor) {
		this.floor = floor;
	}
}

// Test

console.log("\nTest\n")

function test1_requestElevator() {
	column = new Column(10, 2);
	column.arrayElevators[1].currentFloor = 2
	column.arrayElevators[1].direction = null
	column.arrayElevators[1].status = "idle"
	column.arrayElevators[1].queue = []
	column.arrayElevators[0].currentFloor = 6
	column.arrayElevators[0].direction = null
	column.arrayElevators[0].status = "idle"
	column.arrayElevators[0].queue = []
	
	//Calling the function to request an elevator
	column.requestElevator(7, "up");
}
test1_requestElevator();
// //test 2
 
// function test2_requestFloor() {
// 	column2 = new Column(10, 2);
// 	column2.arrayElevators[1].currentFloor = 3
// 	column2.arrayElevators[1].direction = "up"
// 	column2.arrayElevators[1].status = "moving"
// 	column2.arrayElevators[1].queue = [3]
	
// 	column2.requestFloor(7)
// }

// test2_requestFloor()